junit-jupiter-api/src/main/java/org/junit/jupiter/api/Assertions.java

```diff
@@ -1479,12 +1561,97 @@ private static void failArraysNotEqual(Object expected, Object actual, Deque<Int
 		return result;
 	}
 
-	private static String formatIndexes(Deque<Integer> indexes) {
-		if (indexes == null || indexes.isEmpty()) {
-			return "";
+	// --- assertIterableEquals helpers ----------------------------------
+
+	private static void assertIterableEquals(Iterable<?> expected, Iterable<?> actual, Deque<Integer> indexes,
+			Supplier<String> messageSupplier) {
+
+		if (expected == actual) {
+			return;
 		}
-		String indexesString = indexes.stream().map(Object::toString).collect(joining("][", "[", "]"));
-		return " at index " + indexesString;
+		assertIterablesNotNull(expected, actual, indexes, messageSupplier);
+
+		Iterator<?> expectedIterator = expected.iterator();
+		Iterator<?> actualIterator = actual.iterator();
+
+		int processed = 0;
+		while (expectedIterator.hasNext() && actualIterator.hasNext()) {
+			Object expectedElement = expectedIterator.next();
+			Object actualElement = actualIterator.next();
+
+			if (expectedElement == actualElement) {
+				processed++;
+				continue;
+			}
+
+			indexes.addLast(processed++);
```

Comment:
This is neat and smart, but sometimes people oversee the increment within those statements. Would it be okay for you to move the increment on a separate line? We might combine the two increments in lines 1583 and 1587 into one line, maybe as the first statement within the while loop block?